diff --git a/plugin/forward/forward.go b/plugin/forward/forward.go
index 2dc939bc7..603d31043 100644
--- a/plugin/forward/forward.go
+++ b/plugin/forward/forward.go
@@ -8,6 +8,7 @@ import (
 	"context"
 	"crypto/tls"
 	"errors"
+	"strings"
 	"net"
 	"net/http"
 	"slices"
@@ -45,7 +46,7 @@ type Forward struct {
 	hcInterval time.Duration
 
 	from    string
-	ignored []string
+	ignored map[string]struct{}
 
 	nextAlternateRcodes []int
 	nextOnNodata        bool
@@ -81,7 +82,7 @@ type Forward struct {
 
 // New returns a new Forward.
 func New() *Forward {
-	f := &Forward{maxfails: 2, tlsConfig: new(tls.Config), expire: defaultExpire, readTimeout: defaultReadTimeout, p: new(random), from: ".", hcInterval: hcInterval, dohMethod: http.MethodPost, opts: proxyPkg.Options{ForceTCP: false, PreferUDP: false, HCRecursionDesired: true, HCDomain: "."}}
+	f := &Forward{maxfails: 2, tlsConfig: new(tls.Config), expire: defaultExpire, readTimeout: defaultReadTimeout, p: new(random), from: ".", ignored: make(map[string]struct{}), hcInterval: hcInterval, dohMethod: http.MethodPost, opts: proxyPkg.Options{ForceTCP: false, PreferUDP: false, HCRecursionDesired: true, HCDomain: "."}}
 	return f
 }
 
@@ -291,8 +292,10 @@ func (f *Forward) isAllowedDomain(name string) bool {
 		return true
 	}
 
-	for _, ignore := range f.ignored {
-		if plugin.Name(ignore).Matches(name) {
+	ss := strings.Split(name, ".")
+        for i := 0; i < len(ss); i++ {
+                n := strings.Join(ss[i:], ".")
+                if _, ok := f.ignored[n]; ok {
 			return false
 		}
 	}
